### DISCUSSION TITLE
Sample app and underlying machinery - my observations

### DISCUSSION THREAD

**Discussion by Greg-Tichy** [orchetect/MenuBarExtraAccess#1] [General]
My first impression after looking at the code is WOW! I learned a lot how the problems can be solved, far beyond my intuition and imagination 😢. Well, I seriously started to learn Swift/SwiftUI just two months ago... I wonder where is all that knowledge coming from. Apple documentation is, let's say, poor. StackOverflow? Or just Google? Cannot believe, one must first know what to look for. For example I didn't know that window has something like class name, that its descendants can be iterated and so on. 

Working on some app for macOS, that shall live in MenuBar, I found two ways: working directly with (1) NSStatusBar (old school) or use (2) MenuBarExtra. I want to have in StatusItem something more complicated than just an Image, and so I have come to this Project. Thanks Google, of course. I've found an example how to attach a View to the button, dealing with NSStatusItem directly but SwiftUI hides it, as we know. So, I started to play with Demo app accompanying all the machinery. I removed all the StatusItems but the third - and it stopped working. Digging deeper I noticed the Id of StatusItem is used only to distinguish between things created. The question is, which app could possibly want to have many of them? Apples say StatusItem is a scarce thing, many of them is a StatusBar orgy. Maybe I lack the imagination, but cannot guess a proper use of the feature (except, obviously, the Demo app).

This piece of code somehow confirms the idea:

`
    /// Returns the underlying status items created by `MenuBarExtra` for the
    /// `MenuBarExtra` with the specified index.
    ///
    /// Each `MenuBarExtra` creates one status item.
    ///
    /// If the `isInserted` binding on a `MenuBarExtra` is set to false, it may not return a status
    /// item. This may also change its index.
    static func statusItem(for ident: StatusItemIdentity? = nil) -> NSStatusItem? {
        let statusItems = statusItems
        
        guard let ident else { return statusItems.first }
        
        switch ident {
        case .id(let menuBarExtraID):
            return statusItems.filter { $0.menuBarExtraID == menuBarExtraID }.first
        case .index(let index):
            guard statusItems.indices.contains(index) else { return nil }
            return statusItems[index]  // shall return item containing index?
        }
    }
`
The last 3 lines of above code exhibit the dependency on assumed numbering of the items, starting from 0, increment 1. When there's only 3, item will never be found. BTW having only one MenuBarExtra I got **two** status items in the array. Suspicious, isn't it?

Anyhow, I changed Id to 0 and it works like a charm. See screen recording below.

https://user-images.githubusercontent.com/37987808/231454298-9477bb78-f95c-46c1-8236-a81cf022475e.mov

If you are interested in my progress, I will continue posting in this thread.
Once again, chapeaux bas! Your work is highly appreciated.

**Comment by orchetect** [comment]
> I wonder where is all that knowledge coming from.

For this particular library, 95% of it I had to hack out on my own. It took days to find solutions and most of it came from hunches from past experience. As far as NSStatusItem or MenuBarExtra, there isn't much out there apart from Apple's docs. Google is scant to say the least. Which is why I decided to take the deep dive and work out some of the missing functionality.

> I want to have in StatusItem something more complicated than just an Image

That's tricky and may only be possible with NSStatusItem currently. MenuBarExtra is very strict in limiting the menu bar button to being an image, image and text, or text. No custom UI or formatting is allowed.

> I removed all the StatusItems but the third - and it stopped working.

That might happen if you left its index unchanged. The indexes must be supplied starting from 0 in the order in which the MenuBarExtras are listed. If there's only one MenuBarExtra, you can omit the index because it defaults to 0. For two or more MenuBarExtra, you need explicit indexes for them.

As an aside: Within the library, ID and Index are separate things. ID was an experiment that I left in while trying to identify a MenuBarExtra/NSStatusItem instance with a String, but there is no trivial way to achieve it (thus far). So it's not actually used. I decided to settle on Index because it was a little more reliable.

> When there's only 3, item will never be found.

Not sure what you're referring to here. As I mentioned, you have to supply the index. I would love it to be more automatic, but it's the simplest I could get it while maintaining compatibility with multiple MenuBarExtra.

> BTW having only one MenuBarExtra I got two status items in the array. Suspicious, isn't it?

I've never seen this, so not sure what to make of it.

**Reply by Greg-Tichy** [reply]
Image with Text in NSStatusBar can be done:
<img width="435" alt="Screenshot 2023-04-14 at 06 03 08" src="https://user-images.githubusercontent.com/37987808/231938887-dde6b29b-94c2-4624-b42f-d0de57f4d209.png">
From you I got code to get it 😀. Now the question was when to look for it. I tried with onChange and scenePhase but without some window it never fires (and when fires for some other reason there are still no windows).  Last resort is NSApplicationDelegate and when applicationDidFinishLaunching is called all windows are alive and one can style 🤾.

_created here_
`MenuBarExtra("Menu: Index 3", systemImage: "fanblades.fill") {`
_styled here_
`class MyAppDelegate: NSObject, NSApplicationDelegate {
    func applicationWillTerminate(_ aNotification: Notification) {
        // Insert code here to tear down your application
        print("will terminate")
    }
    func applicationDidFinishLaunching(_ aNotification: Notification) {
        print("finished launching with \(NSApp.windows.count) windows")
        if let statusItem = Utils.statusItems.first {
            print("got it!")
            if let button = statusItem.button {
                button.image = NSImage(systemSymbolName: "camera",  accessibilityDescription: "Funny Stuff")
                button.title = "We'll see soon enough"
            }
            
        }
    }
}
`
Remarks:
- I added a new file to Project for AppDelegate and s**t happened
<img width="261" alt="Screenshot 2023-04-14 at 06 24 38" src="https://user-images.githubusercontent.com/37987808/231941422-7a6a3e73-0807-4996-8639-cc689a32969f.png">
quite similar to issue I encountered in another Project of yours. It is there but Xcode refuses to include it in the Project as source file. Weird.

- In an another Project the same operation works as expected.
- 2 NSStatusItem instances for one MenuBarExtra? I have 2 screens. 
- Such a long status item is problematic. On the big screen is visible but on the built-in there's not enough space. Shows nothing.

Concerning index, I think it shall be generated internally and make avail as read-only property, just as you said.

Anyhow, both Projects of yours exhibit this strange behavior. Can you, for example, move ContentView to separate source file and build the Demo App? 

I have now Ventura 13.3.1 and Xcode 14.3, updated yesterday. Should be the same as you have.

**Reply by orchetect** [reply]
> Image with Text in NSStatusBar

You can achieve that without accessing NSStatusItem. Just pass a Label instance into MenuBarExtra that has an Image and Text.

```swift
// separate image and text can be used
MenuBarExtra {
    // your menu content
} label: {
    Image(systemName: "fanblades.fill")
    Text("My text")
}

// or using Label to combine image and text
MenuBarExtra {
    // your menu content
} label: {
    Label("My text", systemImage: "fanblades.fill")
}
```

Perhaps you are trying to do more than that, in which case you may need to access NSStatusItem directly - as I mentioned, MenuBarExtra is strict in only allowing image and/or text. I haven't tried changing the button contents directly on NSStatusItem while using MenuBarExtra. It may work but may not stick if MenuBarExtra is updating itself periodically. At worst it could crash at unexpectex times, as it's not expecting outside interference with its NSStatusItem.

> 2 NSStatusItem instances for one MenuBarExtra? I have 2 screens.

Ah right. Yes I documented this in some comment blocks and I'm just now reminded of it. Technically if you only have one status item but two displays with Spaces on each display, there is only one NSStatusItem and the other is a replica subclass NSStatusItemReplicant:

https://github.com/orchetect/MenuBarExtraAccess/blob/5af5a7d6768862b47d64a51f7343b0a6c81aba86/Sources/MenuBarExtraAccess/MenuBarExtraUtils.swift#L40-L49

**Comment by Greg-Tichy** [comment]
Xcode here doesn't like it
<img width="591" alt="Screenshot 2023-04-14 at 07 41 48" src="https://user-images.githubusercontent.com/37987808/231951646-617141c9-932e-4b69-b947-ceb3e5bc1d01.png">

**Reply by orchetect** [reply]
That's because you still have this.

<img width="591" alt="231951646-617141c9-932e-4b69-b947-ceb3e5bc1d01" src="https://user-images.githubusercontent.com/11605557/231952051-394ef3c7-1733-484b-8e80-f88a0991f2b6.png">

**Comment by Greg-Tichy** [comment]
Right. Silly I.
Still, I want to change text and probably fiddle with icon, maybe change color.

To the right of my thing you can see an another item. It has icon, two lines of text and a tooltip. And is updating all the time.

**Reply by orchetect** [reply]
Some of that is possible with MenuBarExtra. There are some quirks/bugs that prevent you from doing it properly within the `App` scene, but it can be accomplished if you use a custom view with a class that holds state. I'll spare the details but it's a workaround that allows the image and text to change whenever you want. Just access the properties through `menuBarViewModel`.

It can't style anything that MenuBarExtra can't normally style, but it allows changing the status item button contents using SwiftUI.

```swift
@main struct YourApp: App {
    @ObservedObject var menuBarViewModel = MenuBarViewModel()

    var body: some Scene {
        MenuBarExtra {
            // your menu content here
        } label: {
            MenuBarView()
                .environmentObject(menuBarViewModel)
        }
    }
}

struct MenuBarView: View {
    @EnvironmentObject var menuBarViewModel: MenuBarViewModel

    var body: some View {
        Image(systemName: menuBarViewModel.imgName)
        Text(menuBarViewModel.text)
    }
}

class MenuBarViewModel: ObservableObject {
    @Published var imgName: String = "fanblades.fill"
    @Published var text: String = "My Text"
}
```

**Comment by orchetect** [comment]
Just further to this - if you want to get dangerous and poke around at the NSStatusItem, I have pushed changes to `main` branch that give access to it.

See this post for details:

https://github.com/orchetect/MenuBarExtraAccess/discussions/2#discussioncomment-5667588
